src/Features/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs

```diff
@@ -0,0 +1,572 @@
+            Debug.Assert(token.VirtualChars.Length > 0);
+            if (TryMatch(token, "NaN", JsonKind.NaNLiteralToken, out var newKind) ||
+                TryMatch(token, "true", JsonKind.TrueLiteralToken, out newKind) ||
+                TryMatch(token, "null", JsonKind.NullLiteralToken, out newKind) ||
+                TryMatch(token, "false", JsonKind.FalseLiteralToken, out newKind) ||
+                TryMatch(token, "Infinity", JsonKind.InfinityLiteralToken, out newKind) ||
+                TryMatch(token, "undefined", JsonKind.UndefinedLiteralToken, out newKind))
+            {
+                return new JsonLiteralNode(token.With(kind: newKind));
+            }
+
+            if (Matches(token, "-Infinity"))
+            {
+                SplitLiteral(token, out var minusToken, out var newLiteralToken);
+
+                return new JsonNegativeLiteralNode(
+                    minusToken, newLiteralToken.With(kind: JsonKind.InfinityLiteralToken));
+            }
+
+            var firstChar = token.VirtualChars[0];
+            if (firstChar == '-' || firstChar == '.' || IsDigit(firstChar))
+                return ParseNumber(token);
+
+            return new JsonTextNode(
+                token.With(kind: JsonKind.TextToken).AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                    string.Format(FeaturesResources._0_unexpected, firstChar.ToString()),
+                    firstChar.Span)));
+        }
+
+        private JsonConstructorNode ParseConstructor(JsonToken token)
+        {
+            var savedInConstructor = _inConstructor;
+            _inConstructor = true;
+
+            var result = new JsonConstructorNode(
+                token.With(kind: JsonKind.NewKeyword),
+                ConsumeToken(JsonKind.TextToken, FeaturesResources.Name_expected),
+                ConsumeToken(JsonKind.OpenParenToken, s_openParenExpected),
+                ParseSequence(),
+                ConsumeToken(JsonKind.CloseParenToken, s_closeParenExpected));
+
+            _inConstructor = savedInConstructor;
+            return result;
+        }
+
+        private static bool TryMatch(JsonToken token, string val, JsonKind kind, out JsonKind newKind)
+        {
+            if (Matches(token, val))
+            {
+                newKind = kind;
+                return true;
+            }
+
+            newKind = default;
+            return false;
+        }
+
+        private static bool Matches(JsonToken token, string val)
+        {
+            var chars = token.VirtualChars;
+            if (chars.Length != val.Length)
+                return false;
+
+            for (var i = 0; i < val.Length; i++)
+            {
+                if (chars[i] != val[i])
+                    return false;
+            }
+
+            return true;
+        }
+
+        private static bool IsDigit(VirtualChar ch)
+            => ch.Value is >= '0' and <= '9';
+
+        private static JsonLiteralNode ParseNumber(JsonToken textToken)
```

Comment:
`ParseNumber` really hurts my brain for what it's actually doing, which is just creating a literal node from any token.

Reply:
true!  i'll make this better.